maven/rules.bzl

```diff
@@ -187,25 +189,44 @@ def _aggregate_dependency_info_impl(target, ctx):
     deps_all = deps + exports + runtime_deps
 
     maven_coordinates = find_maven_coordinates(target, tags)
-    dependency = None
+    dependencies = []
 
+    # depend via POM
     if maven_coordinates:
-        # depend via POM
-        dependency = struct(
+        dependencies = [struct(
             type = "pom",
             maven_coordinates = maven_coordinates
-        )
-    else:
+        )]
+    # include generic compilation_outputs in JAR
+    elif target[OutputGroupInfo].compilation_outputs:
+        source_jars = target[OutputGroupInfo]._source_jars.to_list()
         # include in the JAR
-        dependency = struct(
+        dependencies = [struct(
             type = "jar",
             class_jar = target[OutputGroupInfo].compilation_outputs.to_list()[0],
-            source_jar = target[OutputGroupInfo]._source_jars.to_list()[-1],
-        )
+            source_jar = source_jars[-1] if source_jars else None,
+        )]
+    # include runtime output jars
+    elif target[JavaInfo].runtime_output_jars:
+        jars = target[JavaInfo].runtime_output_jars
+        source_jars = target[JavaInfo].source_jars
+        dependencies = [struct(
+            type = "jar",
+            class_jar = jar,
+            source_jar = source_jar,
+        ) for (jar, source_jar) in zip(
+            jars, source_jars + [None] * (len(jars) - len(source_jars))
+        )]
 
     return JarInfo(
         name = maven_coordinates,
-        deps = depset([dependency], transitive = [target[JarInfo].deps for target in deps_all]),
+        deps = depset(dependencies, transitive = [
+            # Filter transitive JARs from dependency that has maven coordinates
+            # because those dependencies will already include the JARs as part
+            # of their classpath
+            depset([dep for dep in target[JarInfo].deps.to_list() if dep.type == 'pom'])
+                if target[JarInfo].name else target[JarInfo].deps for target in deps_all
+        ]),
```

Comment:
Can you create a reproducible example to demonstrate this problem? All targets in https://github.com/sugarmanz/bazel-distribution-local-jars appear to build OK without this particular change.

Reply:
> All targets in https://github.com/sugarmanz/bazel-distribution-local-jars appear to build OK without this particular change.

In the sense that Bazel is able to build them successfully, yes. But the desired output is not met. Before this change, `lib-b` would bundle `lib-v` even though it is a transitive dependency that is already bundled and published in `lib-a`:

`lib-a.jar`:
![Screen Shot 2022-03-25 at 1 00 02 PM](https://user-images.githubusercontent.com/9255651/160167637-33fe00ae-75e6-4311-8908-774f82ad1a6b.png)

`lib-b.jar`:
![Screen Shot 2022-03-25 at 12 53 33 PM](https://user-images.githubusercontent.com/9255651/160167538-1728f73d-9a72-4cb1-bf9f-658d417626e6.png)

With this change, it filters out things that will get bundled into the JAR from the transitive dependencies, resulting in `lib-b` no including `lib-v` files as part of the classpath.

`lib-a.jar`:
![Screen Shot 2022-03-25 at 12 53 12 PM](https://user-images.githubusercontent.com/9255651/160167848-c6eb27a5-3e82-420a-9a15-9b513be13100.png)

`lib-b.jar`:
![Screen Shot 2022-03-25 at 12 59 49 PM](https://user-images.githubusercontent.com/9255651/160167878-b70f719c-9c83-4d81-ab36-34df950c0920.png)

Here are the JARs built against commit `bf78b8a` (before that change) of my branch:

https://github.com/sugarmanz/bazel-distribution-local-jars/tree/main/jars/2

Here are the JARs built against latest (`f0be8da`):

https://github.com/sugarmanz/bazel-distribution-local-jars/tree/main/jars/desired

Reply:
Understood. This is a good change.